### PR TITLE
Ability to (de)serialize NetworkConfigBuilder

### LINKED ISSUE #873: Make `NetworkConfig` serializable
### Describe the feature

The idea of the `Config` struct used in the CLI is that it is a thin wrapper around existing config structs to the extent possible to lower maintenance, i.e. instead of having to add new config in 2 places, you just add it in one.

Currently though, it is not possible to serialize or deserialize the `NetworkConfig` struct (or `NetworkConfigBuilder`) because:

1. It takes a generic parameter `C` as a stand-in for the client.
2. It holds a `Box<dyn BlockImport>`

There are a few possible paths here:

1. We try to make `NetworkConfigBuilder` serializable by moving the non-serializable parts into the `build` function, or
2. We make `NetworkConfig` serializable, which means the networking components need to take the unserializable things separately from the config,
3. We make a new slimmer version of `NetworkConfig` that is `Serialize` and `Deserialize`, and we have `NetworkConfig` hold one of these

### Additional context

As part of this issue we would also need to deduplicate a few things. Currently we have a `PeersConfig` as part of the `Config` struct, but this is also present on `NetworkConfig`.

### PR DESCRIPTION
**Pull request by leruaa**
Closes #873

### CODE REVIEW

**Comment by leruaa** [comment]
Still need to make a bunch of things (De)Serializable. Added some `#[serde(skip)]` for now.

**Review by gakonst — COMMENTED** [review]
Looks good so far

**Comment by leruaa** [comment]
Made some progress, some things to notice:

- I've added `skip` on `ban_list` field on `Discv4Config` because `Instant` is not serializable
- `executor` field is still skipped, and I noticed executor() is neved called on the builder. Do I need to better handle this field?
- Same thing for `block_import`, the corresponding method was never called on the builder, So I didn't added the parameter to `build()` and just create a NetworkConfig with `ProofOfStakeBlockImport::default()`. Is that ok?

Thanks in advance for your feedbacks

File: crates/net/nat/src/lib.rs [hunk 1]
```diff
@@ -21,7 +22,7 @@ use std::{
 use tracing::warn;
 
 /// All builtin resolvers.
-#[derive(Debug, Clone, Copy, Eq, PartialEq, Default, Hash)]
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Default, Hash, Serialize, Deserialize)]
```
Comment (mattsse) [inline crates/net/nat/src/lib.rs:25]: I think these should be manually implemented with `FromStr` and `Display`

File: crates/net/network/src/session/config.rs [hunk 2]
```diff
@@ -9,7 +11,7 @@ use std::time::Duration;
 pub const INITIAL_REQUEST_TIMEOUT: Duration = Duration::from_secs(20);
 
 /// Configuration options when creating a [SessionManager](crate::session::SessionManager).
-#[derive(Debug)]
+#[derive(Debug, Serialize, Deserialize)]
```
Comment (mattsse) [inline crates/net/network/src/session/config.rs:14]: serde is optional so this should be

```rust
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
```

File: crates/net/network/src/session/config.rs [hunk 3]
```diff
@@ -55,7 +57,7 @@ impl SessionsConfig {
 /// Limits for sessions.
 ///
 /// By default, no session limits will be enforced
-#[derive(Debug, Clone, Default)]
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
```
Comment (mattsse) [inline crates/net/network/src/session/config.rs:60]: same

File: crates/net/dns/src/tree.rs [hunk 4]
```diff
@@ -202,7 +203,7 @@ impl fmt::Display for BranchEntry {
 }
 
 /// A link entry
-#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+#[derive(Debug, Clone, Hash, Eq, PartialEq, Serialize, Deserialize)]
```
Comment (mattsse) [inline crates/net/dns/src/tree.rs:206]: this should also serialize/deserialize with Display + FromStr

**Review by mattsse — CHANGES_REQUESTED** [review]
ty, 

some early pointers.

File: crates/net/nat/src/lib.rs (hunk 1 above)
Comment (leruaa) [reply crates/net/nat/src/lib.rs:25]: I will use SerializeDisplay and DeserializeFromStr from serde_with if that's ok?

**Comment by gakonst** [comment]
@leruaa this is looking great -just a bunch of merge conflicts to address

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**